tracks/julia/exercises/word-count/mentoring.md

```diff
@@ -0,0 +1,124 @@
+# Example solutions
+
+`````
+
+Docstring:
+
+````julia
+"""
+    wordcount(input)
+
+Find the frequency of each term in the input.
+
+## Example
+
+```jldoctest
+julia> wordcount("Joe can't tell between 'large' and large.")
+Dict{String,Int64} with 6 entries:
+  "tell"    => 1
+  "can't"   => 1
+  "large"   => 2
+  "joe"     => 1
+  "between" => 1
+  "and"     => 1
+```
+"""
+````
+
+**Split, then strip quotation**
+
+This solution, like all the others below uses "Regex".
+Regex is a concise language for describing patterns in strings.
+
+In Julia, `r"foo"` constructs a Regex at compilation time using the `@r_str` macro.
+
+```julia
+function wordcount(input)
+    words = split(input, r"[^\w']+")
+    counts = Dict{String, Int}()
+    for word in words
+        # Removing leading and closing apostrophes (might be quotation)
+        word = lowercase(strip(word, '''))
+        if !isempty(word)
+            counts[word] = get(counts, word, 0) + 1
+        end
+    end
+    return counts
+end
+```
+
+This solution uses a regex, but you could equally do something like:
+
+```julia
+split(input, c -> !(isalnum(c) || c == '''); keepempty=false)
+```
+
+Using `keepempty` because unlike the Regex (which can match many characters at once), this would return many empty strings for long runs of whitespace or punctuation.
+
+**Find each word with regex without splitting the input**
+
+This is a straightforward algorithm and it is easy to prove that it follows the specification exactly.
+Implementations that split by whitespace and punctuation tend to count words that the specification does not ask for, e.g. "x1".
+You can view that as a good thing or a bad thing!
+For what it is worth, traditional word counting code like `wc` treat a mix of letters and digits as a word.
+
+```julia
+function wordcount(input)
+    counts = Dict{String, Int}()
+    # Unicode version matching any letters or numbers: r"\pL+'\pL|\pL+|\pN+"
+    for m in eachmatch(r"[A-Za-z]+'[A-Za-z]|[A-Za-z]+|[0-9]+", input)
```

Comment:
This one fails on contractions like `we're`

```julia
julia> wordcount2("I said 123 and we're done")
Dict{String,Int64} with 7 entries:
  "e"    => 1
  "done" => 1
  "said" => 1
  "123"  => 1
  "we'r" => 1
  "and"  => 1
  "i"    => 1
```

Reply:
Huh. They pass the tests. I should have written my own to check.

Reply:
Oh, there should just be an extra plus. I missed it off when transcribing, I think.

Reply:
Thanks!